### DISCUSSION TITLE
Flux2, SSA, and GitOps

### DISCUSSION THREAD

**Discussion by texasbobs** [fluxcd/flux2#2266] [General] [answered]
Prior to v0.18.0, before SSA, the source was applied to the cluster and any fields that were added to an object from another manager like `kubectl appy` would be removed.  Flux made sure that the cluster was always in the desired state; a major component of GitOps.  This reconciliation happened every time the `interval` period from the kustomization definition passed.  

With Flux2 now taking advantage of SSA, it is now possible for fields to be controlled by managers other than Flux.  This would result in the cluster not matching the desired state.  For example, if someone modifies configuration at the server level and adds a field, it might not be reflected in Github. This results in our source of truth via GitOps doesn't apply anymore and applications might have misconfigurations, issues, and/or cause production outage

How does Flux's use of SSA fit in with the GitOps model given that it can no longer make sure the cluster is in the desired state?

**Comment by stefanprodan** [comment, marked as answer]
> Prior to v0.18.0, before SSA, the source was applied to the cluster and any fields that were added to an object from another manager like kubectl appy would be removed. 

This was never the case with Flux. Before SSA, Flux was using kubectl 3-way-merge, which means that fields added to an object outside of Git, wouldn't be removed unless they overlap with what's specified in Git. For example, if you don't specify `spec.replicas` in a Deployment, then Flux would not override the replicas set by HPA.

With SSA, Flux applies object with [force ownership](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/options.go#L658):
> ForceOwnership indicates that in case of conflicts with server-side apply, the client should acquire ownership of the conflicting field.  Most controllers should use this.

Like before, SSA does override the fields if their in-cluster values differ from what's specified Git. For SSA, there is a [know issue](https://github.com/fluxcd/kustomize-controller/issues/486) for not being able to remove fields in-cluster by removing them from Git if those fields have multiple managers e.g. `kubectl-client-side-apply`.

**Reply by texasbobs** [reply]
Thanks for the reply, Stefan.  I thought the SSA was introducing new situations where the desired state was not going to be obtained.

**Reply by stefanprodan** [reply]
I'm working on the fix for being able to delete fields from Git that have been applied with kubectl, to fully restore the v0.18 behaviour. Going to post a release candidate on that issue.

**Reply by billykwooten** [reply]
@stefanprodan Do you happen to have a link to the PR/branch for that fix so I can track it on my side?

**Reply by stefanprodan** [reply]
Here it is https://github.com/fluxcd/kustomize-controller/pull/527
